lib/iterator.js

```diff
@@ -0,0 +1,312 @@
+'use strict';
+
+const toIterator = base => {
+  if (base[Symbol.iterator]) {
+    return base[Symbol.iterator]();
+  } else if (base.next) {
+    return base;
+  } else {
+    throw new TypeError('Base is neither Iterator nor Iterable');
+  }
+};
+
+class Iterator {
+  constructor(base) {
+    this.base = toIterator(base);
+  }
+
+  [Symbol.iterator]() {
+    return this;
+  }
+
+  next() {
+    return this.base.next();
+  }
+
+  each(fn, thisArg) {
+    this.forEach(fn, thisArg);
+  }
+
+  forEach(fn, thisArg) {
+    for (const value of this) {
+      fn.call(thisArg, value);
+    }
+  }
+
+  every(predicate, thisArg) {
+    for (const value of this) {
+      if (!predicate.call(thisArg, value)) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  find(predicate, thisArg) {
+    for (const value of this) {
+      if (predicate.call(thisArg, value)) {
+        return value;
+      }
+    }
+  }
+
+  includes(element) {
+    for (const value of this) {
+      if (value === element) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  reduce(reducer, initialValue) {
+    let result = initialValue;
+
+    if (result === undefined) {
+      const next = this.next();
+      if (next.done) {
+        throw new TypeError(
+          'reduce of consumed iterator with no initial value'
+        );
+      }
+      result = next.value;
+    }
+
+    for (const value of this) {
+      result = reducer(result, value);
+    }
+    return result;
+  }
+
+  some(predicate, thisArg) {
+    for (const value of this) {
+      if (predicate.call(thisArg, value)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  collectTo(CollectionClass) {
+    return new CollectionClass(this);
+  }
+
+  collectWith(obj, collector) {
+    this.forEach(element => collector(element, obj));
+  }
+
+  toArray() {
+    return Array.from(this);
+  }
+
+  map(mapper, thisArg) {
+    // eslint-disable-next-line no-use-before-define
+    return new MapIterator(this, mapper, thisArg);
+  }
+
+  filter(predicate, thisArg) {
+    // eslint-disable-next-line no-use-before-define
+    return new FilterIterator(this, predicate, thisArg);
+  }
+
+  flat(depth = 1) {
+    // eslint-disable-next-line no-use-before-define
+    return new FlatIterator(this, depth);
+  }
+
+  flatMap(mapper, thisArg) {
```

Comment:
This function throws on this basic valid example:
```js
common.iter([1, 2, 3]).flatMap(x => x*2).toArray()
TypeError: Base is neither Iterator nor Iterable
    at toIterator (common/lib/iterator.js:9:11)
    at new Iterator (common/lib/iterator.js:15:17)
    at FlatMapIterator.next (common/lib/iterator.js:230:30)
    at Function.from (native)
    at FlatMapIterator.toArray (common/lib/iterator.js:99:18)
```

Reply:
Fixed.